snuba/admin/clickhouse/system_queries.py

```diff
@@ -129,13 +112,62 @@ def _run_sql_query_on_host(
 )
 
 
-def is_query_select(sql_query: str) -> bool:
+def is_query_using_only_system_tables(
+    clickhouse_host: str,
+    clickhouse_port: int,
+    storage_name: str,
+    sql_query: str,
+) -> bool:
     """
-    Simple validation to ensure query is a select command
+    Run the EXPLAIN QUERY TREE on the given sql_query and check that the only tables
+    in the query are system tables.
     """
-    sql_query = " ".join(sql_query.split())
-    match = SYSTEM_QUERY_RE.match(sql_query)
-    return True if match else False
+    sql_query = sql_query[:-1] if sql_query.endswith(";") else sql_query
+    explain_query_tree_query = (
+        f"EXPLAIN QUERY TREE {sql_query} SETTINGS allow_experimental_analyzer = 1"
+    )
+    explain_query_tree_result = _run_sql_query_on_host(
+        clickhouse_host, clickhouse_port, storage_name, explain_query_tree_query, False
+    )
+
+    for line in explain_query_tree_result.results:
+        line = line[0].strip()
+        # We don't allow table functions for now as the clickhouse analyzer isn't good enough yet to resolve those tables
+        if line.startswith("TABLE_FUNCTION"):
+            return False
+        if line.startswith("TABLE"):
+            match = re.search(r"table_name:\s*(\S+)", line, re.IGNORECASE)
+            if match:
+                table_name = match.group(1)
+                if not table_name.startswith("system."):
+                    return False
+
+    return True
+
+
+def is_valid_system_query(
+    clickhouse_host: str, clickhouse_port: int, storage_name: str, sql_query: str
+) -> bool:
+    """
+    Validation based on Query Tree and AST to ensure the query is a valid select query.
+    """
+    explain_ast_query = f"EXPLAIN AST {sql_query}"
+    disallowed_ast_nodes = ["AlterQuery", "AlterCommand", "DropQuery", "InsertQuery"]
```

Comment:
Would it be better to use allowed nodes rather than disallowed nodes? I worry if a new node type is ever made it could be an issue.

Reply:
This disallowed node check is just an extra layer of protection, we already use a readonly mode when querying clickhouse which provides the main level of protection. If I made a list of allowed nodes, it would be much longer.